File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/Serializer.java
package it.unibo.coordination.linda.presentation;

import java.io.Writer;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public interface Serializer<T> {

    MIMETypes getSupportedMIMEType();

    Object toDynamicObject(T object);

    default List<Object> toDynamicObject(Collection<? extends T> objects) {
        return objects.stream().map(this::toDynamicObject).collect(Collectors.toList());
    }

    String toString(T object);

    String toString(Collection<? extends T> objects);

    void write(T object, Writer writer);

    void write(Collection<? extends T> objects, Writer writer);

    static <X> Serializer<X> of(Class<X> klass, MIMETypes mimeType) {
        return Presentation.getSerializer(klass, mimeType);
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/Serializer.kt
package it.unibo.presentation

import java.io.Writer

interface Serializer<T> {
    val supportedMIMEType: MIMETypes

    fun toDynamicObject(`object`: T): Any

    @JvmDefault
    fun toDynamicObject(objects: Collection<T>): List<Any> {
        return objects.asSequence()
                .map { `object`: T -> this.toDynamicObject(`object`) }
                .toList()
    }

    fun toString(`object`: T): String

    fun toString(objects: Collection<T>): String

    fun write(`object`: T, writer: Writer)

    fun write(objects: Collection<T>, writer: Writer)
}